source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -124,6 +141,36 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
   return {action, bytes_read, end_stream};
 }
 
+void SslSocket::complete() {
+  ASSERT(async_handshake_in_progress_);
+  async_handshake_in_progress_ = false;
+
+  ENVOY_CONN_LOG(debug, "async handshake complete", callbacks_->connection());
+  if (!handshake_complete_) {
+    // It's possible that the async call comes in later, but the handshake has been retried from
+    // doWrite or similar. In that case the error handling has been done synchronously. */
+    PostIoAction result = doHandshake();
+    if (async_handshake_in_progress_ || result != PostIoAction::KeepOpen) {
+      // It may be semantically possible that the handshake is expecting read() or write(), so
+      // we shouldn't require that the handshake is complete. However, there can't be a second
+      // private key operation requested.
+      if (callbacks_->connection().state() == Network::Connection::State::Open) {
```

Comment:
This code is a bit tricky, can you talk about how this is tested and whether there is a way to simplify? Would definitely want to hear from @PiotrSikora on this.

Reply:
Yes, this is quite complex. The idea here is that we can't leave handshakes hanging. Either they need to be completed or they need to error out. We only get the error result when `ssl_do_handshake()` is called the second time and the internal completion function in private key method is run. As you say, this function is quite crucial because the other end of the SSL connection can partially control the handshake and connection status. A bug in the logic could result in the connection never being closed or successful, leaving Envoy resources tied. @PiotrSikora, do you have any comments or opinions here?

If the second call to `ssl_do_handshake()` returns an error, we need to tell Envoy that the handshake isn't going anywhere, and this is why `callbacks_->connection().close()` is called. The problem with this is that this results in a `LocalClose` event, but synchronous failing results in `RemoteClose` event. It seems that this can't be easily fixed without adding a new parameter to `close()` call, which I didn't want to do to keep the PR scope more limited.

Testing the callback function is not very straightforward, because it depends on internal SSL connection state and the Envoy connection state machine. The current tests for success and failure cases are in `ssl_socket_test.cc`. All the cases which expect a `LocalClose` event are testing the error cases, for example this: https://github.com/envoyproxy/envoy/pull/6326/files#diff-be9d0f766b258e0940a6d9b1d35a7878R4325. Maybe the ssl_socket test harness could be extended to have better control over client connection state.